### ISSUE TITLE
[yul] optimizer's load resolver does not take side effects of user-defined function containing for loop(s) into account

### ISSUE THREAD

**Issue by bshastry** [ethereum/solidity#8032]
## Description

```
{
        function foo(x) {
                // codecopy alters contents of mem[0]
                for {} x { x := mload(0) codecopy(0, 77, 32)} {}
        }
        // mem[0] = 1337
        mstore(0, 1337)
        // Function with side-effects that alters memory
        foo(42)
        // mem[0] = 1337 may no longer be valid here
        sstore(0, mload(0))
}
```

is optimized to

```
            let _1 := 1337      
            mstore(0, _1)       
            let x := 42                                                
            for { }             
            x                   
            {       
                x := mload(0)
                codecopy(0, 77, 32)
            }
            { }
            // mload(0) is incorrectly resolved to initial assignment
            sstore(0, _1)
        }
```

## Environment

- Compiler version: latest develop

## Steps to Reproduce

```
$ solc --strict-assembly --optimize <test.yul>
```

**Comment by chriseth** [comment]
@bshastry is the `codecopy` essential here?

**Comment by bshastry** [comment]
> @bshastry is the `codecopy` essential here?

I think so because I've not been able to alter contents of memory in other ways that the load resolver does not see so far. May be I'm missing some trick to do it without codecopy.

**Comment by bshastry** [comment]
I'm wrong. codecopy is not essential.

```
{
        function foo(x) {
                for {} x { x := mload(0) mstore(0, 0)} {}
        }
        mstore(0, 1337)
        foo(42)
        sstore(0, mload(0))
}
```

**Comment by bshastry** [comment]
This bug affects 0.5.14 as well.

**Comment by chriseth** [comment]
Fixed in #8034
